File: apps/android/app/src/main/java/com/schedulersystems/scheduler/ui/screens/export/SchedulePdfRenderer.kt
package com.schedulersystems.scheduler.ui.screens.export

import android.content.Context
import android.graphics.Color
import android.graphics.Paint
import android.graphics.pdf.PdfDocument
import android.graphics.pdf.PdfRenderer
import android.os.ParcelFileDescriptor
import com.schedulersystems.scheduler.domain.export.SchedulePdfDoc
import java.io.File

/**
 * Result of rendering a [SchedulePdfDoc] to a file: the file plus the page count
 * read back from the actual written PDF (proof it is a valid, non-empty document
 * rather than trusting our own pagination loop).
 */
data class RenderedPdf(val file: File, val pageCount: Int)

/**
 * Renders a [SchedulePdfDoc] to an A4 PDF using android.graphics.pdf.PdfDocument —
 * a real, on-device PDF (no stub, no external service). Lives in the UI layer
 * because it depends on android.graphics; the pure row/title/filename logic is in
 * domain/export/SchedulePdf.kt and is unit-tested separately.
 */
object SchedulePdfRenderer {
    private const val PAGE_W = 595 // A4 @72dpi, points
    private const val PAGE_H = 842
    private const val MARGIN = 40f
    private const val ROW_H = 22f

    fun render(context: Context, doc: SchedulePdfDoc): RenderedPdf {
        val pdf = PdfDocument()
        val titlePaint = Paint().apply { color = Color.BLACK; textSize = 16f; isFakeBoldText = true }
        val headerPaint = Paint().apply { color = Color.rgb(168, 85, 247); textSize = 11f; isFakeBoldText = true }
        val bodyPaint = Paint().apply { color = Color.BLACK; textSize = 10f }
        val footerPaint = Paint().apply { color = Color.GRAY; textSize = 8f }

        val colX = floatArrayOf(MARGIN, MARGIN + 110f, MARGIN + 220f)
        val rowsPerPage = ((PAGE_H - 130) / ROW_H).toInt()
        // Always at least one page, even when the roster is empty.
        val chunks = if (doc.rows.isEmpty()) listOf(emptyList()) else doc.rows.chunked(rowsPerPage)

        chunks.forEachIndexed { pageIdx, chunk ->
            val info = PdfDocument.PageInfo.Builder(PAGE_W, PAGE_H, pageIdx + 1).create()
            val page = pdf.startPage(info)
            val canvas = page.canvas
            var y = 50f
            if (pageIdx == 0) {
                canvas.drawText(doc.title, MARGIN, y, titlePaint)
                y += 30f
            }
            SchedulePdfDoc.HEADER.forEachIndexed { i, h -> canvas.drawText(h, colX[i], y, headerPaint) }
            y += ROW_H
            chunk.forEach { r ->
                canvas.drawText(r.shift, colX[0], y, bodyPaint)
                canvas.drawText(r.day, colX[1], y, bodyPaint)
                canvas.drawText(r.employee, colX[2], y, bodyPaint)
                y += ROW_H
            }
            canvas.drawText("Page ${pageIdx + 1} of ${chunks.size}", PAGE_W - 110f, PAGE_H - 20f, footerPaint)
            pdf.finishPage(page)
        }

        val file = File(context.cacheDir, doc.filename)
        file.outputStream().use { pdf.writeTo(it) }
        pdf.close()

        // Reopen the written file and count pages from the real bytes.
        val pageCount = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY).use { pfd ->
            PdfRenderer(pfd).use { it.pageCount }
        }
        return RenderedPdf(file, pageCount)
    }
}

File: apps/ios/Sources/SchedulerApp/Services/SchedulePdfRenderer.swift
import UIKit

/// Result of rendering a SchedulePdfDoc: the written file URL plus the page count
/// read back from the real rendered bytes (proof it is a valid, non-empty PDF).
struct RenderedSchedulePdf {
    let url: URL
    let pageCount: Int
}

/// Renders a SchedulePdfDoc to an A4 PDF using UIGraphicsPDFRenderer — a real,
/// on-device PDF (no stub, no external service). Mirrors the Android
/// SchedulePdfRenderer; the pure row/title/filename logic lives in SchedulePdf.swift
/// and is unit-tested separately.
enum SchedulePdfRenderer {
    private static let pageWidth: CGFloat = 595 // A4 @72dpi, points
    private static let pageHeight: CGFloat = 842
    private static let margin: CGFloat = 40
    private static let rowHeight: CGFloat = 22

    static func render(_ doc: SchedulePdfDoc) -> RenderedSchedulePdf {
        let pageRect = CGRect(x: 0, y: 0, width: pageWidth, height: pageHeight)
        let renderer = UIGraphicsPDFRenderer(bounds: pageRect)

        let rowsPerPage = max(1, Int((pageHeight - 130) / rowHeight))
        let chunks: [[SchedulePdfRow]] = doc.rows.isEmpty
            ? [[]]
            : stride(from: 0, to: doc.rows.count, by: rowsPerPage).map {
                Array(doc.rows[$0..<min($0 + rowsPerPage, doc.rows.count)])
            }

        let colX: [CGFloat] = [margin, margin + 110, margin + 220]
        let titleAttr: [NSAttributedString.Key: Any] = [.font: UIFont.boldSystemFont(ofSize: 16)]
        let headerAttr: [NSAttributedString.Key: Any] = [
            .font: UIFont.boldSystemFont(ofSize: 11),
            .foregroundColor: UIColor(red: 168 / 255, green: 85 / 255, blue: 247 / 255, alpha: 1)
        ]
        let bodyAttr: [NSAttributedString.Key: Any] = [.font: UIFont.systemFont(ofSize: 10)]
        let footerAttr: [NSAttributedString.Key: Any] = [
            .font: UIFont.systemFont(ofSize: 8), .foregroundColor: UIColor.gray
        ]

        let data = renderer.pdfData { ctx in
            for (pageIdx, chunk) in chunks.enumerated() {
                ctx.beginPage()
                var y: CGFloat = 50
                if pageIdx == 0 {
                    (doc.title as NSString).draw(at: CGPoint(x: margin, y: y), withAttributes: titleAttr)
                    y += 30
                }
                for (i, header) in SchedulePdfDoc.header.enumerated() {
                    (header as NSString).draw(at: CGPoint(x: colX[i], y: y), withAttributes: headerAttr)
                }
                y += rowHeight
                for row in chunk {
                    (row.shift as NSString).draw(at: CGPoint(x: colX[0], y: y), withAttributes: bodyAttr)
                    (row.day as NSString).draw(at: CGPoint(x: colX[1], y: y), withAttributes: bodyAttr)
                    (row.employee as NSString).draw(at: CGPoint(x: colX[2], y: y), withAttributes: bodyAttr)
                    y += rowHeight
                }
                let footer = "Page \(pageIdx + 1) of \(chunks.count)"
                (footer as NSString).draw(at: CGPoint(x: pageWidth - 110, y: pageHeight - 30), withAttributes: footerAttr)
            }
        }

        let url = FileManager.default.temporaryDirectory.appendingPathComponent(doc.filename)
        try? data.write(to: url)

        let pageCount: Int = {
            guard let provider = CGDataProvider(data: data as CFData),
                  let pdf = CGPDFDocument(provider) else { return 0 }
            return pdf.numberOfPages
        }()

        return RenderedSchedulePdf(url: url, pageCount: pageCount)
    }
}
